### PR TITLE
Enable MFA in WingedKeys

### PR DESCRIPTION
**Pull request by JordanGuinn**
## Background
This PR adds support for multi-factor authentication in our WingedKeys application.

## GitHub Issue
- https://github.com/ctoec/winged-keys/issues/75

## Associated PRs
N/A

## Validation Plan
- [x] Confirm user can use ECE Reporter after confirming MFA token
- [x] Confirm user **cannot** use ECE Reporter without confirming MFA token
- [x] Confirm user cannot use ECE Reporter after specifying invalid MFA token
- [x] Confirm user creation through WingedKeys UI enables MFA by default
- [x] Confirm navigation to the MFA confirmation view does not send a confirmation code email to the user
- [x] Confirm users **without** MFA-enabled can still login simply with username and password

## Additional Context
As part of this work, the `TwoFactorEnabled` column will need to be manually updated to `1` for all users that we want to apply this to.  Otherwise, login will continue as it did previously.

### CODE REVIEW

**Review by bamader — APPROVED** [review]
I _think_  I understand what's going on here? Idk, the `AccountController` is pretty nasty. My biggest question is why is this all structured as an MVC pattern? Doesn't seem like the most intuitive way to process an authentication.

**Comment by JordanGuinn** [comment]
WingedKeys is still a web app at the end of the day, so the MVC pattern here doesn't seem all that far-fetched imo.  I also imagine it was the easiest/quickest way to get this thing spun up when we first started way back when 😅

File: src/WingedKeys/Quickstart/Account/AccountController.cs [hunk 1]
```diff
@@ -108,37 +110,16 @@ public async Task<IActionResult> Login(LoginInputModel model, string button)
...
-                    }
-
-                    // request for a local page
-                    if (Url.IsLocalUrl(model.ReturnUrl))
-                    {
-                        return Redirect(model.ReturnUrl);
-                    }
-                    else if (string.IsNullOrEmpty(model.ReturnUrl))
+                    return await RedirectAfterLogin(context, model.Username, model.ReturnUrl);
+                } else if (result.RequiresTwoFactor)
```
Comment (vonbearshark) [inline src/WingedKeys/Quickstart/Account/AccountController.cs:114]: wouldn't we want the `RequiresTwoFactor` as the first check?

File: src/WingedKeys/Quickstart/Account/AccountController.cs [hunk 2]
```diff
@@ -151,7 +132,49 @@ public async Task<IActionResult> Login(LoginInputModel model, string button)
...
+            return View(new TwoFactorViewModel());
+        }
+
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public async Task<IActionResult> ConfirmTwoFactorToken(TwoFactorInputModel twoFactorModel, string returnUrl)
+        {
+            if (!ModelState.IsValid)
+            {
+                return View(twoFactorModel);
```
Comment (vonbearshark) [inline src/WingedKeys/Quickstart/Account/AccountController.cs:149]: Should we `Error =` here as well?

File: src/WingedKeys/Quickstart/Account/AccountController.cs [hunk 3]
```diff
@@ -151,7 +132,49 @@ public async Task<IActionResult> Login(LoginInputModel model, string button)
...
+            if (user == null)
+            {
+                return RedirectToAction(nameof(Login));
+            }
+            var result = await _signInManager.TwoFactorSignInAsync("Email", twoFactorModel.TwoFactorCode, false, false);
+            if (result.Succeeded)
+            {
+                return await RedirectAfterLogin(context, user.UserName, returnUrl);
+            }
+            else if (result.IsLockedOut)
```
Comment (vonbearshark) [inline src/WingedKeys/Quickstart/Account/AccountController.cs:165]: wouldn't we want this to supersede `Succeeded`?

**Review by vonbearshark — APPROVED** [review]
Happy to press more if needed but seems reasonable enough

File: src/WingedKeys/Quickstart/Account/AccountController.cs (hunk 3 above)
Comment (JordanGuinn) [reply src/WingedKeys/Quickstart/Account/AccountController.cs:165]: Another instance where it doesn't actually matter - in practice a `SignInResult` will never have both succeeded and also have locked us out

**Comment by nickbristow** [comment]
Validation:
- [ ] Confirm user can use ECE Reporter after confirming MFA token
- [ ] Confirm user cannot use ECE Reporter without confirming MFA token
- [ ] Confirm user cannot use ECE Reporter after specifying invalid MFA token
- [ ] Confirm user creation through WingedKeys UI enables MFA by default
- [ ] Confirm navigation to the MFA confirmation view does not send a confirmation code email to the user
- [ ] Confirm users without MFA-enabled can still login simply with username and password

All steps validated. Good to go.

**Comment by JordanGuinn** [comment]
WOOOOT - will move to `Ready`, and hold off on merge/deploy until all communications have been sent out and we're prepared to push the button.

**Comment by JordanGuinn** [comment]
Merging so that we can finally close this ticket - we can follow-up on how we want to go about handling release at another moment, but the actual dev work here is done.

**Pull request merged**